### DISCUSSION TITLE
Offgridtec LiFePo4 Smart-Pro

### DISCUSSION THREAD

**Discussion by patman15** [fl4p/batmon-ha#84] [Ideas]
Hi @fl4p,
I have implemented a working addition to your repository, so I can readout the [Offgridtec LiFePo4 Smart-Pro series](https://www.offgridtec.com/offgridtec-lifepo4-smart-pro-12-30-akku-12-8v-384wh-lithium-batterie.html) of batteries. Unfortunately, I have no idea which BMS is used inside the battery (and I do not want to disassemble it) and I also do not know how to figure it out. Nevertheless, readout works, but the data is scrambled by some key derived from the serial number.
Long story short, in case you are interested, I can spend some more hours to figure out how this is done so my implementation can be used with any battery and not only mine.

![HA_2023-04-16](https://user-images.githubusercontent.com/14628713/232872861-b9102147-5271-43e0-aa08-45fd70a8c9e1.png)

Btw. I also found a few minor issues in the main class (BtBms) in case you want some more thoughts. :)

**Comment by fl4p** [comment]
Hi @patman15 ,
definitely interested in adding this to batmon.
Also let me know what you found in `BtBms`

Cheers
Fabian

**Reply by patman15** [reply]
Hi @fl4p,

> definitely interested in adding this to batmon.

Ok, I will take a bit of time, I need to figure out how the general algorithm for the data scambling works. :-\

> Also let me know what you found in `BtBms`
>
`BtBms`: Shouldn't that class by default read the [BT device info](https://www.bluetooth.com/specifications/specs/device-information-service-1-1/)? That could give all devices a unique name and probably some additional information.

For the `class BmsSample` I have troubles with the capacity/SoC calculations together with my BMS. I'm not sure how the different parameters `charge, capacity, cycle_capacity` are defined by you. In the end, my BMS delivers a SoC, a factory capacity (what is written on the battery) and a current charge in Ah.
If I provide `capacity` (factory) and `charge` (current contained energy), the `SoC` in percent is "wrong" as it is calculated against the factory capacity and not the `cycle_capacity` (which I assume it the currently available full charge value which is lower than the factory capacity). see also https://www.biologic.net/topics/battery-states-state-of-charge-soc-state-of-health-soh/
If you look at
https://github.com/fl4p/batmon-ha/blob/01fabc01ef808f03b59a5584c5d1a3d0e9ea478b/bmslib/bms.py#L54-L55
I guess the line should read
`if capacity > 0 and ((math.isnan(soc) or isinstance(soc, int)) and charge > 0):` (note the brackets)
as otherwise if you provide no `soc` and no `charge`, `SoC` will run into the assert.

I hope I remembered the issues correctly from my whole tryouts. I'll come back with a PR for my BMS as soon as I have figured out the general solution for descrambling.

**Reply by fl4p** [reply]
I missed your reply.
You're right aber the soc computation and the assert. I removed the assert in the meanwhile.

This is my definition of charge, capacity and cycle_capacity:
```
        :param charge: The charge available in Ah, aka remaining capacity, between 0 and `capacity`
        :param capacity: The capacity of the battery in Ah
        :param cycle_capacity: Total absolute charge meter (coulomb counter). Can tell you the battery cycles (num_cycles = cycle_capacity/2/capacity). A better name would be cycle_charge. This is not well defined.
        
        
```

cycle_capacity increases when both charging and discharging.

**Reply by patman15** [reply]
Ok, thanks, then I got it right.

**Comment by gkathan** [comment]
hi,
I am currently trying to get my offgridtec smart pro Akku from my van integrated to my HA instance in the van ... is your work already in the latest batmon build ? would love to use it => why battery type do I have to choose if it is already in latest build ?
cheerz from vienna

**Reply by patman15** [reply]
> hi, I am currently trying to get my offgridtec smart pro Akku from my van integrated to my HA instance in the van ... is your work already in the latest batmon build ?

No it is not yet integrated. I couldn't (yet) figure out how the scrambling is related to the serial number. Thus, I cannot provide a general solution that would easily work for everyone.

> would love to use it => why battery type do I have to choose if it is already in latest build ? 

You need to have a battery with a serial number Bxxxxx then we probably can make it work by trial and error (16 possibilities).

> cheerz from vienna

Same here!

**Reply by fl4p** [reply]
@patman15 If its only 16 possibilities, we can just brute force it. You want to create a pull-request so we can integrate it into upstream? Doesn't need to be perfect and 100% tested.

**Reply by patman15** [reply]
Hmm, good point, so far I thought I could figure out the dependency from the serial number (as there is some), but actually you are right. Ok, then I need to generate the missing 15 descramble patterns. Need to find some time for that, but I will start with solving @gkathan issue than it's only 14 to go. ;-)

**Reply by gkathan** [reply]
nice :-) what do you mean with "14 to go" ? ... how long do you think will this take you ? i am eager to get my battery connected in HA :-)

**Reply by patman15** [reply]
There are 16 different scrambled communication ways. 1 I solved with my battery and one could be yours. Then there are 14 left.

**Comment by gkathan** [comment]
there is a WEEE registration number printed on the battery ... is that wahr
we need?

**Reply by patman15** [reply]
Nope, it's the number that shows also in the app. You should read something like "SmartPro-Bxxxxx" the 5 "x" digits are the relevant ones. At least my battery also has a silver sticker on top with a (suppose) manufacturing date and then "Bxxxxx" as well.
If you provide me the number, I can generate a "fake" battery using an Arduino device and then disassemble the scrambling for your case, by monitoring the connection between the app and the "fake" battery.

**Reply by gkathan** [reply]
ok....the smart akku app does not show the detail page in my case ... already reported this to offgridtec ... will search for a serialnumber on the battery itself tonight (it is quite hidden in my van...)

**Reply by patman15** [reply]
... and what if you press the "scanning" button in the app? Then it should list you the battery, including it's number.

**Reply by gkathan** [reply]
![PXL_20230822_081422780](https://github.com/fl4p/batmon-ha/assets/6329814/48748fca-3a34-4ca7-8434-cb4792d02b39)
so....found the sticker on the battery with the serialnumber (checked with offgridtec) 
it is:

20210514A16380

...but looking at that ... seems like the manufacturing date 
hmmmm

**Reply by gkathan** [reply]
unfortunately the app shows only SOC, voltage and the detail drill-in is empty

**Reply by patman15** [reply]
Yep, that was the number I was looking for A16380. Unfortunately, this is an "A" type battery and not the "B" I have and I know that the "A" has different scrambling and different functionality. We would need to intercept communication and figure out how that works. That is a longer project ...

**Comment by gkathan** [comment]
ok ... once I have fixed the BMS of the battery I can fiddle around with
intercepting the traffic ... but it looks like the BMS somehow died in the
course of my yesterday's full charging cycle ... in contact with offgridtec
already ... looks like I have to file a warranty case and return the
battery ... which sucks as I need it in my van ...

**Reply by patman15** [reply]
Did you try disconnecting it (completely) for some minutes and then to charge it slightly? That helped for my battery.

**Comment by gkathan** [comment]
i was told that I can try to completly discharge it which causes the BMS to reset / reboot .... cannot do this right now because i just did such a cycle which caused the problem ;-) looks like it is somehow broken ... will buy another 200Ah battery ... and once this arrives will send back the broken one ... and when it is fixed use it as second (2x200Ah)
will tell you once new battery is here for testing (curious whether it is an "A" or "B" model

**Reply by gkathan** [reply]
hi,
got a new 200Ah battery of same type....what shall I do to sniff the Bluetooth protocol?

**Reply by gkathan** [reply]
...is worth buying such a sniffer ? https://www.antratek.de/bluefruit-le-sniffer-bluetooth-low-energy-ble-4-0-nrf51822?gclid=Cj0KCQjwi7GnBhDXARIsAFLvH4nSqcKhBjXnV8WopCg6yoKl1N6S-W6oOKMZPXgueVaRzfUWKbCDbAAaAp7UEALw_wcB

**Reply by gkathan** [reply]
or is it "classic" Bluetooth (not low energy) ... Ubertooth One 2.4GHz Wireless Development Bluetooth Sniffer BTLE Hacking Tool Bluetooth Protocol Analysis A https://amzn.eu/d/7nRRN4q

**Reply by fl4p** [reply]
you don't need to buy hardware. which OS do you use? on iOS , macos you can use PaketLogger.

also see https://github.com/fl4p/batmon-ha/blob/master/doc/BT%20Sniffing.md

**Comment by gkathan** [comment]
have android, Linux and windows ... no iOS

**Reply by fl4p** [reply]
for android see this https://stackoverflow.com/questions/23877761/sniffing-logging-your-own-android-bluetooth-traffic

if you have a app on linux or windows that connects to the bms you can use Wireshark

**Comment by gkathan** [comment]
ok ... and for what shall i look for once the sniffer is active ?

**Reply by fl4p** [reply]
It depends on how the BMS BLE interface works. Some BMS just need an initial subscribe command from the app and they will periodically send binary data with the updated battery info. Others only respond to a request, so the app will need to frequently send poll requests.

Can't give you a general guide on how to reverse engineer this. I think best is to start with decompiling the BMS app to get an idea about the binary format of the data frames

**Reply by gkathan** [reply]
ok
any preferred APK decompiler ? will try apk-tool otherwise and have orderer the BLE sniffer from adafruit to also take a look on real-time traffic... have a new lipo4 battery at home for some days to play around (also hacking into the victron connect system to enable HA access to those guys;-)
cheerz

**Reply by fl4p** [reply]
https://www.decompiler.com/
usually it is easier to do the sniffing on the software side, because that doesn't involve encryption.
capturing bluetooth packets is the easy part. understanding them is hard

**Reply by patman15** [reply]
@gkathan I can tell you how I did it, but again, please note, I have a "B" type of battery which probably has a different BMS.
I used BT debugging on my phone to record the communication between the battery and the app.
Then I wrote a sketch for an ESP32 to emulate a "fake" battery that would reply to BLE request to check which values are shown in the app correspond to the requests. Did a lot of value debugging and trial and error. Finally I wrote an ESP32 sketch to query the battery and show the numbers. Last I transformed the latter into a class for batmon.
I didn't know how to decompile the apk, but that is probably an easier approach. Let me know if you need something for my approach or if you need my sample code (ESP32 or the class for batmon). I can somewhere upload it but as mentioned, currently it's of limited use for users of batmon ...

**Reply by gkathan** [reply]
...successfully decompiled the offgridtec smart Akku Viewer APK  and now studying the source code to find the relevant snippets ... will let you know

**Reply by gkathan** [reply]
> @gkathan I can tell you how I did it, but again, please note, I have a "B" type of battery which probably has a different BMS.
> I used BT debugging on my phone to record the communication between the battery and the app.
> Then I wrote a sketch for an ESP32 to emulate a "fake" battery that would reply to BLE request to check which values are shown in the app correspond to the requests. Did a lot of value debugging and trial and error. Finally I wrote an ESP32 sketch to query the battery and show the numbers. Last I transformed the latter into a class for batmon.
> I didn't know how to decompile the apk, but that is probably an easier approach. Let me know if you need something for my approach or if you need my sample code (ESP32 or the class for batmon). I can somewhere upload it but as mentioned, currently it's of limited use for users of batmon ...

is it possible to get access to your code ... would like to learn how to access the offgridtec battery ...
can you drop me your python code off your working offgridtec type"B" Plugin ?

**Reply by gkathan** [reply]
hi,
can you let me look at your code / implementation of the "B" type offgridtec akku ... so that I can derive / learn how to access the "A" battery ? 
cheerz and thanx

**Reply by patman15** [reply]
Sure! I'll fork and upload the changes their. Give me a few days to cleanup and prepare.

**Comment by gkathan** [comment]
would like to understand better how to write a class for batmon ... is
reading the documentation of batmon sufficient to get an understanding how
to code it ?

**Reply by patman15** [reply]
@gkathan I now created a [fork](https://github.com/patman15/batmon-ha) of the current version (needed to integrate some updates). On the branch [OGT](https://github.com/patman15/batmon-ha/tree/ogt/) you can find my additions that allow me access to my battery. Just have a look at the [BMS class](https://github.com/patman15/batmon-ha/blob/ogt/bmslib/models/ogt.py) and note that the (de)scrambling of the messages is in a [separate class](https://github.com/patman15/batmon-ha/blob/ogt/bmslib/models/conv.py). The string currently only works for my serial number (and most likely every 16th other).
Wish you success, let me know if it was helpful or I can help more.

**Reply by gkathan** [reply]
thanx so much,
will try to gets my hands dirty in coding once back in Vienna ... what is the smartest way to setup a local HA add-on development environment ? did the docker setup but realized that the add-ons are not there with that method ... 
cheerz

**Reply by patman15** [reply]
Probably @fl4p has better experience than me, I simply threw the git repository into the _addon_ folder and installed it as local addon. That way I can simply edit the files and then rebuild to test. I'm pretty sure there are more sophisticated methods than that. 8-)

**Reply by fl4p** [reply]
You can just run the add-on stand-alone, that's how I usually do it. Install the requirements, and create `options.json` where you put the BMS address. see [doc/Standalone.md](https://github.com/fl4p/batmon-ha/blob/master/doc/Standalone.md).
it'll work without mqtt set up, just see the command line logs

**Reply by patman15** [reply]
@gkathan just had a look into `/sources/com/yscoco/smartbattery/utils/EncryptUtils.java` which seems to be the code for this scambling stuff. Looks like the _A_ type batteries just differ in the way the `EncyKey` is derived from the serial number. Which would mean, with the right key you should be able to access your battery with the rest of my code if you can figure out which registers are available from `/sources/com/yscoco/smartbattery/activity/DeviceInfoActivitiy.java`

**Reply by gkathan** [reply]
ok ... now found some time to read your code ... some questions:
* the   _CRYPT = b'\x29\x28\x2B\x2A\x2D\x2C\x2F\x2E\x21\x20\x58\x5B\x5A\x5D\x5C\x5F' is coming from where ? is this the same "key" when you refer to the type A battery encryption key which is derived from my serial number ?
* I cannot find the conv.py file in your oft branch ... am I blind ;-) 
* so you mean with just tweaking / understanding the encryption from the decompiled java classes ... all I need to do is to define the "correct" _CRYPT byte array ?
thanx

**Reply by patman15** [reply]
>     * the   _CRYPT = b'\x29\x28\x2B\x2A\x2D\x2C\x2F\x2E\x21\x20\x58\x5B\x5A\x5D\x5C\x5F' is coming from where ? is this the same "key" when you refer to the type A battery encryption key which is derived from my serial number ?
> 
Yes.

>     * I cannot find the conv.py file in your oft branch ... am I blind ;-)
> 
Did you use the links in my first post / made sure you are on the OGT branch?

>     * so you mean with just tweaking / understanding the encryption from the decompiled java classes ... all I need to do is to define the "correct" _CRYPT byte array ?
>
That would be my assumption for the ideal case.

**Comment by gkathan** [comment]
hi,
just found some time to suceesfully get my DEV environment together and tried to get your fork/branch with ogt support running ...  currently i get in the logs the following:

looks like this error indicates that something fundamental is missing on my side
exec: fatal: unable to exec bashio
: No such file or directory

Batmon
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
exec: fatal: unable to exec bashio
: No such file or directory
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped

**Comment by gkathan** [comment]
ok ... somehow i came a step further ... have now your branch/fork running on my raspi in the van where my "a" type battery is running ... current log shows that encryption has to be adapted ... now i can try to hack this :-)

**Reply by patman15** [reply]
@gkathan I think I finally figured out how to derive the encryption from the serial number -- not tested yet. Further, I wrote a native BT integration for HA that automatically detects my battery and adds it. See below
![screenshot](https://github.com/fl4p/batmon-ha/assets/14628713/66e1451f-e2ce-4118-a072-a520f9c9fa45)

@fl4p I'm not sure that this is also what you were looking for as well? It's still quite beta, but I'd like to make an intermediate layer to use the models from [/bmsliib/models](https://github.com/fl4p/batmon-ha/tree/master/bmslib/models) to be able to integrate various BT batteries. Not sure if this is possible with the current interfaces. Let me know if working without MQTT is something that you had in mind when writing `TODO: use the new Bluetooth integration since HA 2022.8?` because then I could help now.

**Reply by gkathan** [reply]
you are just plain awesome ✨✨ .. yez ... I am looking for exactly this ... possible that I can get access to try ? ... my van is still in repair but I hope for not so Long anymore
cheeerz from vienna

**Reply by 7wells** [reply]
@patman15 
> Further, I wrote a native BT integration for HA that automatically detects my battery and adds it. 

Would you mind sharing details about this? Thank you so much! 👍

**Reply by patman15** [reply]
@7wells There is a screenshot above what else would you like to know?

**Reply by fl4p** [reply]
@patman15 yes, porting batmon to the HA integration API is my intention, haven't started it yet. 
let's try to join our efforts here.
I haven't dig much into the integration API.
It would probably make most sense to have a separate repository that can be added to HACS. 
And having everything under `bmslib/` in a git submodule or a pip module, that can be used by both the integration and the add-on.
What are your thoughts?

**Reply by patman15** [reply]
@fl4p thanks for answering!

> let's try to join our efforts here.

Yep, I'd love to!

> I haven't dig much into the integration API.

Well, I'm done, it's working for my BLE battery. My current state is, battery is automatically added as soon as it appears on the BT adapter. As sensors I currently only add those types that are available native in HA. (Could be changed, but I wanted to follow KISS principle ;-)) I'd really like to keep it minimal and do the rest extra. So it depends on how you want to handle, e.g. energy meters (which I would see outside, as there are solutions for that)

> It would probably make most sense to have a separate repository that can be added to HACS.

Agree, I'm just preparing one (will take a few days more) with an initial starting point that works with the OGT battery of mine to make gkathan happy. :-)

> And having everything under bmslib/ in a git submodule or a pip module, that can be used by both the integration and the add-on.

Yes, I do agree to that. Submodule can be done as long as it is HACS, if you think even on going to HA directly, then a pip module would be required. I'm good with both, just noticed that there are still only very few basic BLE integrations in core. Wonder why?
My current integration expects a very simple API. It calls the submodule constructor with an `BLEdevice`. Then only one method is required `async def update(self) -> dict`. The return value is expected to have the format `{"sensor_key": value}`. A sample would be: `{'temperature': 21.0, 'voltage': 13.229, 'current': 0.0, 'battery_level': 56, 'cycle_capacity': 219.73369, 'runtime': 3932100}`
So the only thing left is refactor the classes under `bmslib/` to that interface and allow the integration to detect and instantiate different type (which is simple).

@fl4p should we start a new discussion on that? My plan is to publish the HACS integration on my github account so everyone can have a look. I'm especially interested which functionality of your current solution you expect to be there later. E.g. MQTT, or stand-alone working etc. Probably, aligning on the (interface of the) submodules in `bms-lib/` and re-using them for the two work-flows is even better?
Let me know your thoughts how to take it from there and how I can support you. Excited to join forces!

**Reply by fl4p** [reply]
@patman15 
Yes, let's start a new discussion, its getting a bit messy here otherwise.

I have little time at the moment for batmon.
I think best is to use an intermediate interface, like batmon already does with e.g. `BmsSample`. We can then publish data of this data structure to HA via MQTT (add-on) or the integration API. This way we have a modular approach and we split the sources (BMS) from the data sinks (MQTT, HA integration API)

I will get back as soon as I have more time

**Comment by patman15** [comment]
> possible that I can get access to try ?

Sure, let me refactor a bit and then I will try to publish my first HACS integration. Just wanted to give you some heads-up that things are progressing (even with long voids in between)

**Reply by patman15** [reply]
Unfortunately noticed that the "A" type batteries not only have a different encryption but also a different frame format. Need to have a look again. "B" type does work now.

**Comment by 7wells** [comment]
> @7wells There is a screenshot above what else would you like to know?
Ah, sorry, I might have misunderstood your "wrote a native BT integration for HA that automatically detects my battery and adds it". At least from your screenshot I do not understand what you meant by "native BT integration...".

**Reply by patman15** [reply]
@7wells In contrast to the batmon _addon_, I did an _integration_ that uses HA bluetooth stack and does not require MQTT (addon). Does this answer your question?

**Reply by 7wells** [reply]
Sorry that I might have overlooked the obvious, but do you offer your integration code somewhere?

**Reply by patman15** [reply]
Give me some days to prepare a repository for a HACS integration and then I will publish. Just finished testing today, batteries of type A and B are working now.

**Comment by patman15** [comment]
@7wells, @gkathan I now published my current beta version here: https://github.com/patman15/BLE_BMS-HA feedback welcome.

**Reply by gkathan** [reply]
awesome 🤘🤘 I have to wait till my van is back from service / repair ... then I will test and provide feedback
great work 🌈

**Reply by gkathan** [reply]
soooo ... finally i have my van back and went for a first test-trip ... will check your integration the coming days and will provide feedback 
cheerz![PXL_20240329_134049360.MP.jpg](https://github.com/fl4p/batmon-ha/assets/6329814/22adcbef-3e69-4686-8f6c-923c8fdc6678)

**Reply by patman15** [reply]
Just merged the latest and greatest changes. Please verify that you have version tag `5a29ef2`.
Wish you a relaxing weekend, weather is supposed to be nice. :smile:

**Reply by gkathan** [reply]
hej .... you are just PLAIN AWESOME 🤘🤘🤘 .. just reinstalled my HA raspi in the van and did looots of upgrades first, installed your HACS integration ... AND ... it just WORKED out of the box without a single glitch ... detected by offgridtec battery and all values are online now
fantastic work
and best greetings from Austria
c![Screenshot_20240330-081311.png](https://github.com/fl4p/batmon-ha/assets/6329814/b7aedc3d-c39e-47bf-b34a-e6fcc45b3761)

**Reply by patman15** [reply]
@gkathan thanks for the nice feedback and I'm happy it works! Was quite some effort for 2 users and a product that is not sold anymore. 😁 
Greeting from Austria as well!

**Reply by gkathan** [reply]
you also from Austria ? where ? I live in Vienna and currently in Steyr
* we should meet - the only 2 customers 😂😂

**Reply by fl4p** [reply]
So cool to see your integration working @patman15 !

**Comment by gkathan** [comment]
ok

one question ... it looks like since I connected HA with your integration
to the smartbat ... the offgridtec app no longer sees the battery to
connect to it ... looks like only one connection to the bms is allowed ...
true?
cheerz

gerold kathan :: +43 699 1973536

**Reply by patman15** [reply]
Hey! Correct, the BMS allows only one connection at a time. In order to keep it logic, I decided to permanently keep the connection open within the integration. There is the possibility to change that (integration tries to connect, reads data, disconnects every 30s) but this would mean that it is "random" when the app gets data and when the integration. You can disable the device then it should disconnect after some time and the app work again.
patcalc@buerotiger.de if you want to contact me per PM.

**Comment by patman15** [comment]
@gkathan thanks again for testing the integration! If there are further issues please raise one for [my integration](https://github.com/patman15/BLE_BMS-HA/issues) or contact me directly to not mess up things here.
I will wait for Fabian to come back to me to continue discussions on joining efforts in https://github.com/fl4p/batmon-ha/discussions/213 and close this topic now.

**Reply by gkathan** [reply]
one last thing:
* the only thing missing compared to the app is to have an indication when the battery is charging (boolean and/or the amount of amp which is going into the battery)
* or am I just missing it?
cheeerz

**Reply by patman15** [reply]
Hi! Current sensor should be available, showing negative values for supply and positive values for charging. If you are looking for a separate binary sensor stating charging/non-charging I would need to implement it. Nevertheless, this would be a "sign indicator" of the current, as the BMS does not deliver charging indication natively.
It's even in your picture here https://github.com/fl4p/batmon-ha/discussions/84#discussioncomment-8957790, 2nd from top `current`

**Comment by gkathan** [comment]
hi
just created an issue in the repo
* the Current and Power entities/sensors are both always zero in my case
regardless whether it is charging or discharging

https://github.com/patman15/BLE_BMS-HA/issues/2

if you need Support for the type-A battery please let me know how I can
assist

cheeerz and thänx so much




gerold kathan :: +43 699 1973536
